Add Watergate Sonic Local Integration

adam-the-hero opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Sonic Wi-Fi by Watergate will have a Local API included in the newest firmware, so we are brining this integration to Home Assistant.
Integration uses: https://pypi.org/project/watergate-local-api/


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/35535

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

adam-the-hero committed:
Add Watergate Sonic Local Integration

zweckj commented:
What the bot said: You can only add one platform with the initial PR. Please reduce the PR to one platform and add the other platform in follow up PRs.

zweckj commented on homeassistant/components/watergate/valve.py:
```diff
@@ -0,0 +1,106 @@
+    ) -> None:
+        """Initialize the sensor."""
+        super().__init__(coordinator)
+        self._attr_unique_id = f"{entry.data[SONIC_NAME]}.{VALVE_ENTITY_NAME}"
```
would also be good for a base entity

zweckj commented on homeassistant/components/watergate/valve.py:
```diff
@@ -0,0 +1,106 @@
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle data update."""
+        self._valve_state = (
```
if you're device can go offline an `available` property would make sense (maybe also in base entity)

adam-the-hero replied:
I do not think it will be in base entity. I will put it here for now.

zweckj replied:
your other entities won't have the same logic for their availability?

adam-the-hero replied:
Well, they will have, but the coordinator will fetch a few endpoints, and different entities will need to check if different parts of Aggregated Results are available.
For example for networking I will have data in ```coordinator.data.networking``` not in ```coordinator.data.state```.

zweckj commented on homeassistant/components/watergate/valve.py:
```diff
@@ -0,0 +1,106 @@
+        """Return if the valve is closing or not."""
+        return self._valve_state == "closing"
+
+    def update(self, valve_state: str) -> None:
+        """Update the valve."""
+        self._valve_state = valve_state
+        self.async_write_ha_state()
```
```suggestion
```
no need for this, you got a coordinator

adam-the-hero replied:
I am using it as a webhook handler.

zweckj replied:
could we give this another name then? also annotate with `@callback`

adam-the-hero replied:
Done.

zweckj commented on homeassistant/generated/integrations.json:
```diff
@@ -6871,6 +6871,12 @@
     },
+    "watergate": {
+      "name": "Watergate",
+      "integration_type": "hub",
```
```suggestion
      "integration_type": "device",
```
looks like this integration only provides a single devices at a time

adam-the-hero commented on homeassistant/components/watergate/coordinator.py:
```diff
@@ -0,0 +1,44 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class WatergateAgregatedRequests:
+    """Class to hold aggregated requests."""
+
+    def __init__(self, state: DeviceState | None = None) -> None:
+        """Initialize aggregated requests."""
+        self.state = state
```
As you remember, I've already prepared this coordinator for the rest of the data that we are going to fetch. I will keep it since I will be building on it in the near future.

joostlek replied:
I am missing the point why we should do this? We work incrementally and thus should only have code that is required for the valve. And then this indeed is a wrapper

adam-the-hero commented on homeassistant/components/watergate/strings.json:
```diff
@@ -0,0 +1,29 @@
+          "ip_address": "[%key:common::config_flow::data::ip%]"
+        },
+        "title": "Configure Watergate Device",
+        "description": "Please put name and IP address of your Watergate device"
```
What do you mean bu that?

adam-the-hero replied:
Nvm. Got it.

adam-the-hero commented on homeassistant/components/watergate/valve.py:
```diff
@@ -0,0 +1,106 @@
+        super().__init__(coordinator)
+        self._attr_unique_id = f"{entry.data[SONIC_NAME]}.{VALVE_ENTITY_NAME}"
+        self._entry = entry
+        self._attr_name = VALVE_SENSOR_NAME
```
It is a main feature. I've just removed it.

adam-the-hero commented on homeassistant/components/watergate/config_flow.py:
```diff
@@ -0,0 +1,68 @@
+        self, user_input: dict[str, str] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a flow initiated by the user."""
+        if user_input is not None:
```
Well, I can just check the device state. Added.

zweckj commented on homeassistant/components/watergate/config_flow.py:
```diff
@@ -0,0 +1,73 @@
+        ):
+            return self.async_create_entry(
+                title=user_input[CONF_NAME],
+                data={
+                    CONF_NAME: user_input[CONF_NAME],
+                    CONF_IP_ADDRESS: user_input[CONF_IP_ADDRESS],
+                },
```
```suggestion
                data=user_input
```

zweckj commented on homeassistant/components/watergate/config_flow.py:
```diff
@@ -0,0 +1,73 @@
+        self, user_input: dict[str, str] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a flow initiated by the user."""
+        if user_input is not None and await self.validate_connection(
```
if ``validate_connection` evaluates to `false`, we should show an error (by adding to the `errors` param ofs how_form)

zweckj commented on homeassistant/components/watergate/config_flow.py:
```diff
@@ -0,0 +1,73 @@
+        if user_input is not None:
+            return self.async_update_reload_and_abort(
+                current_entry,
+                data_updates={
+                    CONF_NAME: user_input[CONF_NAME],
+                    CONF_IP_ADDRESS: user_input[CONF_IP_ADDRESS],
+                },
```
```suggestion
                data_updates=user_input,
```

zweckj commented on homeassistant/components/watergate/entity.py:
```diff
@@ -0,0 +1,56 @@
+        self._api_client = coordinator.api
+        self._attr_unique_id = f"{entry.data[CONF_NAME]}.{entity_name}"
+        self._entry = entry
+        self._device_info = DeviceInfo(
+            identifiers={(DOMAIN, entry.entry_id)},
+            name=entry.data[CONF_NAME],
+            manufacturer=MANUFACTURER,
+            sw_version=coordinator.data.state.firmware_version
+            if coordinator.data.state
+            else "",
+        )
+        self._api_client = entry.runtime_data.client
+        entry.runtime_data.entities[entity_name] = self
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return device information for this entity."""
+        return self._device_info
```
```suggestion
        self._attr_device_info = DeviceInfo(
            identifiers={(DOMAIN, entry.entry_id)},
            name=entry.data[CONF_NAME],
            manufacturer=MANUFACTURER,
            sw_version=coordinator.data.state.firmware_version
            if coordinator.data.state
            else "",
        )
        self._api_client = entry.runtime_data.client
        entry.runtime_data.entities[entity_name] = self
```
no need to use the property, you can directly write the attr

zweckj commented on homeassistant/components/watergate/__init__.py:
```diff
@@ -0,0 +1,82 @@
+
+    if (entity := hass.config_entries.async_get_entry(webhook_id)) is not None:
+        if body_type == WEBHOOK_VALVE_TYPE:
+            entity.runtime_data.entities[VALVE_ENTITY_NAME].webhook_update(data.state)
```
why is this necessary? You have one data dict for everything anyways, can't we call `async_set_updated_data` from the coordinator (which will inform all the entities)

adam-the-hero replied:
I guess it looks better now. Thanks for all those suggestions. I wrote integration once, 2 years ago, everything is different now :)

zweckj commented on tests/components/watergate/test_config_flow.py:
```diff
@@ -0,0 +1,173 @@
+        assert result["step_id"] == "user"
+        assert CONF_NAME in result["data_schema"].schema
+        assert CONF_IP_ADDRESS in result["data_schema"].schema
+
```
you can just continue this test here and basically merge the two tests. We want all tests to reach terminal states (entry creation or aborts) anyways

zweckj commented on tests/components/watergate/test_config_flow.py:
```diff
@@ -0,0 +1,173 @@
+
+
+@pytest.mark.asyncio
+async def test_step_reconfigure_update_entry(hass: HomeAssistant) -> None:
```
same here can be one imo

zweckj commented on tests/components/watergate/test_config_flow.py:
```diff
@@ -0,0 +1,173 @@
+            CONF_NAME: "Updated Device",
+            CONF_IP_ADDRESS: "192.168.1.102",
+        },
+    )
```
you'll now also need a test for the error case during validation (you'll need 100% coverage for the configflow)

zweckj commented on tests/components/watergate/test_init.py:
```diff
@@ -0,0 +1,116 @@
+from tests.common import MockConfigEntry
+
+
+@pytest.fixture
```
move those to conftest.py

zweckj commented on tests/components/watergate/test_coordinator.py:
```diff
@@ -0,0 +1,68 @@
+from homeassistant.helpers.update_coordinator import UpdateFailed
+
+
+@pytest.fixture
```
should also be in conftest

zweckj commented on tests/components/watergate/test_coordinator.py:
```diff
@@ -0,0 +1,68 @@
+    return client
+
+
+@pytest.mark.asyncio
```
testing internals (aka directly testing the coordinator) is something that's not liked. Rather test the effects of those things on your entities. The first test would be checked by a valve test, and for the second you can check that the valve become unavailable correctly.

adam-the-hero replied:
Understood. Can you, please show me how it is the best way to do it? I am struggling with a way how to check for example if valve callbacks are executed.

zweckj replied:
which callbacks do you mean? `_handle_coordinator_update` has been correctly called if you verify that the valve has a correct state (which can for example easily be done through `snapshot` tests). The other callbacks you can verify by just calling the webhooks (for that you can check tedee `test_init` or withings again).

adam-the-hero replied:
Ok, sorted. Somehow.

adam-the-hero commented on homeassistant/components/watergate/valve.py:
```diff
@@ -0,0 +1,106 @@
+class SonicValve(CoordinatorEntity[WatergateDataCoordinator], ValveEntity):  # pylint: disable=hass-enforce-class-module
+    """Define a Sonic Valve entity."""
+
+    _attr_has_entity_name = True
```
Done.

adam-the-hero commented on tests/components/watergate/test_config_flow.py:
```diff
@@ -0,0 +1,173 @@
+        # Check that the form is returned with expected fields
+        assert result["type"] == data_entry_flow.FlowResultType.FORM
+        assert result["step_id"] == "user"
+        assert CONF_NAME in result["data_schema"].schema
+        assert CONF_IP_ADDRESS in result["data_schema"].schema
```
This is only showing a form so result["data"] is empty here.

zweckj commented on homeassistant/components/watergate/__init__.py:
```diff
@@ -0,0 +1,103 @@
+
+    _LOGGER.debug("Registered webhook: %s", webhook_id)
+
+    await coordinator.async_refresh()
```
```suggestion
    await coordinator.async_config_entry_first_refresh()
```

zweckj commented on homeassistant/components/watergate/__init__.py:
```diff
@@ -0,0 +1,103 @@
+    await coordinator.async_refresh()
+
+    webhook_url = async_generate_url(
+        hass, webhook_id, allow_external=False, allow_ip=True
```
is there a reason why you are blocking external?

adam-the-hero replied:
No.

zweckj commented on tests/components/watergate/test_init.py:
```diff
@@ -0,0 +1,94 @@
+        await hass.config_entries.async_setup(mock_entry.entry_id)
+        await hass.async_block_till_done()  # Wait for all setup tasks to complete
+
+        assert mock_entry.state == ConfigEntryState.LOADED
```
```suggestion
        assert mock_entry.state is ConfigEntryState.LOADED
```

zweckj commented on tests/components/watergate/test_init.py:
```diff
@@ -0,0 +1,94 @@
+            "homeassistant.components.watergate.WatergateLocalApiClient"
+        ) as mock_client,
+        patch(
+            "homeassistant.components.watergate.WatergateDataCoordinator"
+        ) as mock_coordinator,
```
normally we don't patch the whole coordinator, but only your library inside the coordinator

zweckj commented on tests/components/watergate/test_init.py:
```diff
@@ -0,0 +1,94 @@
+async def test_handle_webhook(hass: HomeAssistant, mock_entry) -> None:
+    """Test handling webhook events."""
+    webhook_id = mock_entry.entry_id
+    hass.config.components.add("network")
+    hass.config.internal_url = "http://hassio.local"
```
what do you need that for?

zweckj commented on homeassistant/components/watergate/const.py:
```diff
@@ -0,0 +1,11 @@
+
+MANUFACTURER = "Watergate AI"
+
+VALVE_SENSOR_NAME = "Valve"
```
appears unused

zweckj commented on homeassistant/components/watergate/entity.py:
```diff
@@ -0,0 +1,52 @@
+            else "",
+        )
+        self._api_client = entry.runtime_data.client
+        entry.runtime_data.entities[entity_name] = self
```
I don't believe you are accessing this anymore, so this can be removed from the `type`

zweckj commented on homeassistant/components/watergate/entity.py:
```diff
@@ -0,0 +1,52 @@
+        super().__init__(coordinator)
+        self._attr_has_entity_name = True
+        self._api_client = coordinator.api
+        self._attr_unique_id = f"{entry.data[CONF_NAME]}.{entity_name}"
```
will be unique across platforms anyways, no need to add the entity name here

adam-the-hero replied:
I am not sure about that. What if someone will have two devices with the same name? This CONF_NAME is something that the user is setting up.

adam-the-hero replied:
Ok. It works. But I am not sure what will happen if we will add more entities. I will have to figure it later.

adam-the-hero replied:
Yes, I got it. This is what I did, but still, I am concerned about what will happen in the future. After this PR, I have more ;) Currently, it works perfectly so there is no need to think about the near future. I will refactor it later.

zweckj replied:
CONF_NAME is not good here. If you have no other ways to generate a unique id here (e.g. serial number of the device), `entry.entry_id` is the fallback option to make sure it's unique

zweckj replied:
Also refactoring this is not really possible since this you would mean a breaking change, so let's get this right now

adam-the-hero replied:
Ok, let me know if you think something else needs to be changed. I do not have anything specific for this device. So, I guess entry_id should be here.

adam-the-hero replied:
But, should I use the entity_name or just entry_id? Entry_id will not be unique across all my later entities, right?

adam-the-hero replied:
Yes, I was passing only valve constant since this is the only entity I have. I cannot have more than one platform. But I will have 20 other entities for that device, so I guess I should revert that entity_name passing, right?

zweckj replied:
Yeah apologies

adam-the-hero replied:
Done. Hopefully, it is everything.

zweckj commented on homeassistant/components/watergate/entity.py:
```diff
@@ -0,0 +1,52 @@
+        self,
+        coordinator: WatergateDataCoordinator,
+        entry: WatergateConfigEntry,
+        entity_name: str,
```
and if you implement the two former suggestions you can remove this param

zweckj commented on homeassistant/components/watergate/const.py:
```diff
@@ -0,0 +1,11 @@
+
+VALVE_SENSOR_NAME = "Valve"
+
+VALVE_ENTITY_NAME = "valve"
```
and then we should be able to get rid of this as well

zweckj commented on homeassistant/components/watergate/__init__.py:
```diff
@@ -0,0 +1,103 @@
+        body_type = body.get("type")
+
+        coordinator_data = coordinator.data
+        if body_type == WEBHOOK_VALVE_TYPE and coordinator_data.state:
```
```suggestion
        if body_type == Platform.VALVE and coordinator_data.state:
```
it's the same string anyways

zweckj commented on tests/components/watergate/conftest.py:
```diff
@@ -0,0 +1,91 @@
+        yield mock_client_instance
+
+
+@pytest.fixture
```
I don't understand why we need to different mock_config_entries tbh?

zweckj commented on tests/components/watergate/test_coordinator.py:
```diff
@@ -0,0 +1,37 @@
+from homeassistant.core import HomeAssistant
+
+
+async def test_change_valve_state(
```
move this to test_valve.py and I'd suggest you throw a snapshot test in there

adam-the-hero replied:
I've added one snapshot assertion. Not for all states. To be honest, in the entity, we do not have anything more than the state, so one is enough for me, the rest of the state changes I am testing with other assertions.

zweckj commented on tests/components/watergate/test_init.py:
```diff
@@ -0,0 +1,81 @@
+    mock_watergate_client,
+) -> None:
+    """Test handling webhook events."""
+    hass.config.components.add("network")
+    await async_setup_component(hass, "http", {"http": {}})
+    await async_setup_component(hass, webhook.DOMAIN, {})
```
those should be unnecessary imo

zweckj commented on tests/components/watergate/test_init.py:
```diff
@@ -0,0 +1,81 @@
+    hass: HomeAssistant, mock_entry, mock_watergate_client
+) -> None:
+    """Test setting up the Watergate integration."""
+    hass.config.components.add("network")
+    hass.config.internal_url = "http://hassio.local"
```
I don't think you need that either (not sure though)

adam-the-hero replied:
It is only to have a little bit prettier assertions instead of some 10.10.10.10 in line 
```
mock_watergate_client.async_set_webhook_url.assert_called_once_with(
            "http://hassio.local/api/webhook/webhook_id"
        )
```

zweckj commented on tests/components/watergate/test_init.py:
```diff
@@ -0,0 +1,81 @@
+
+    with (
+        patch(
+            "homeassistant.helpers.network.get_url", return_value="http://hassio.local"
```
I don' t remember, but doesn't this return a mocked URL by default?

adam-the-hero replied:
It is not required anymore. I am not using this method anymore, and I am configuring it a few lines earlier with the internal_url, so now we have beautiful integration tests, almost E2E :)

zweckj commented on tests/components/watergate/test_coordinator.py:
```diff
@@ -0,0 +1,37 @@
+        blocking=True,
+    )
+
+    assert hass.states.get("valve.sonic").state == "opening"
```
same here

zweckj reviewed (commented):
we are getting there

adam-the-hero commented on homeassistant/components/watergate/config_flow.py:
```diff
@@ -0,0 +1,82 @@
+        except Exception:  # noqa: BLE001
+            return False
+
+    async def async_step_reconfigure(
```
I just used it so many times to have it done and working properly. Tested and so on. Can we just leave here? 🥺

zweckj replied:
Not my decision, but the guys from core who will need to give the final approval will very likely request to remove it 🤷‍♂️ But you can try

zweckj commented on homeassistant/components/watergate/config_flow.py:
```diff
@@ -0,0 +1,49 @@
+        self, user_input: dict[str, str] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a flow initiated by the user."""
+        errors = {}
```
```suggestion
        errors: dict[str, str] = {}
```

zweckj commented on homeassistant/components/watergate/coordinator.py:
```diff
@@ -0,0 +1,44 @@
+            update_interval=timedelta(minutes=2),
+        )
+        self.api = api
+        self.hass = hass
```
```suggestion
```
already done in `super.__init__`

adam-the-hero replied:
True.

zweckj commented on tests/components/watergate/test_config_flow.py:
```diff
@@ -0,0 +1,87 @@
+    hass: HomeAssistant, mock_watergate_client, mock_webhook_id_generatio
+) -> None:
+    """Test checking if registration form works end to end."""
+    flow = WatergateConfigFlow()
+    flow.hass = hass
+
+    result = await flow.async_step_user(user_input=None)
```
```suggestion
       result = await hass.config_entries.flow.async_init(
            DOMAIN, context={"source": SOURCE_USER}
        )
```
is how this is normally done

zweckj commented on tests/components/watergate/test_config_flow.py:
```diff
@@ -0,0 +1,87 @@
+
+    assert result["type"] is data_entry_flow.FlowResultType.CREATE_ENTRY
+    assert result["title"] == "Test Device"
+    assert result["data"][CONF_NAME] == user_input[CONF_NAME]
+    assert result["data"][CONF_IP_ADDRESS] == user_input[CONF_IP_ADDRESS]
+    assert result["data"][CONF_WEBHOOK_ID] == MOCK_WEBHOOK_ID
```
```suggestion
    assert result["data"] == {**user_input, MOCK_WEBHOOK_ID}
```

zweckj commented on tests/components/watergate/test_config_flow.py:
```diff
@@ -0,0 +1,87 @@
+    hass: HomeAssistant, mock_watergate_client_none_result
+) -> None:
+    """Test checking if errors will be displayed when None value is returned while checking device state."""
+    flow = WatergateConfigFlow()
```
same here

adam-the-hero replied:
@zweckj same what? I do not get it.

zweckj commented on tests/components/watergate/test_config_flow.py:
```diff
@@ -0,0 +1,87 @@
+    assert CONF_NAME in result["data_schema"].schema
+    assert CONF_IP_ADDRESS in result["data_schema"].schema
+
+    user_input = {
```
save that in a var to avoid duplicate code

adam-the-hero replied:
I've moved this to the conftest.

zweckj commented on tests/components/watergate/test_config_flow.py:
```diff
@@ -0,0 +1,87 @@
+
+    assert result["type"] is data_entry_flow.FlowResultType.FORM
+    assert result["step_id"] == "user"
+    assert CONF_NAME in result["data_schema"].schema
+    assert CONF_IP_ADDRESS in result["data_schema"].schema
```
don't think that adds value we'll check the result anyways, you definitely don't need that in every test

adam-the-hero replied:
I will remove it in the first one. I just want to be sure nobody changes those keys.

zweckj commented on tests/components/watergate/test_config_flow.py:
```diff
@@ -0,0 +1,87 @@
+
+    result = await flow.async_step_user(user_input=None)
+
+    assert result["type"] is data_entry_flow.FlowResultType.FORM
```
```suggestion
    assert result["type"] is.FlowResultType.FORM
```
preference, but please import from `data_entry_flow`

zweckj commented on tests/components/watergate/test_config_flow.py:
```diff
@@ -0,0 +1,87 @@
+    assert result["errors"][CONF_IP_ADDRESS] == "cannot_connect"
+
+
+async def test_step_user_form_with_exception(
```
you can probably unify those last two tests with parametrization. Also, at the end resolve the error and show the flow can finish successfully (recover from errors)

adam-the-hero replied:
I cannot resolve errors since I am using a different mocked Watergate client to mock exceptions and other wrong responses. But I am checking everything in the previous test.

adam-the-hero replied:
Parametrised.

zweckj replied:
why not? at the end of this test continue to set `mock_watergate_client.async_get_device_state = DeviceState(...)` and you should be able to finish the test with a success

adam-the-hero replied:
Yeah got it now.

zweckj commented on tests/components/watergate/test_init.py:
```diff
@@ -0,0 +1,71 @@
+
+    entity_id = "valve.sonic"
+
+    assert hass.states.get(entity_id).state == "open"
```
assert state first, and use enums

adam-the-hero replied:
What do you mean by 'assert state first'?

adam-the-hero replied:
I know it already.

zweckj commented on tests/components/watergate/test_init.py:
```diff
@@ -0,0 +1,71 @@
+
+    await hass.async_block_till_done()  # Ensure the webhook is processed
+
+    assert hass.states.get(entity_id).state == "closed"
```
idem

zweckj commented on tests/components/watergate/test_init.py:
```diff
@@ -0,0 +1,71 @@
+    """Test handling webhook events."""
+    hass.config.components.add("network")
+
+    mock_entry.add_to_hass(hass)
+
+    await hass.config_entries.async_setup(mock_entry.entry_id)
+    await hass.async_block_till_done()  # Wait for all setup tasks to complete
```
this can be put in a method, since you're using this multiple times

zweckj commented on tests/components/watergate/test_valve.py:
```diff
@@ -0,0 +1,49 @@
+        blocking=True,
+    )
+
+    current_state = hass.states.get(entity_id).state
+    assert current_state
+    assert current_state == ValveState.CLOSING
```
```suggestion
    current_state = hass.states.get(entity_id)
    assert current_state
    assert current_state.state == ValveState.CLOSING
```
for the other occurrences as well. This is to ensure, you'll at least have the entity, before getting it's `state`

adam-the-hero replied:
Got it. But I will do it only once in one test on first occurence, than I will be checking only test.

zweckj reviewed (changes requested):
I think we're pretty good in the normal code now, some more remarks for the tests

adam-the-hero commented on tests/components/watergate/test_valve.py:
```diff
@@ -0,0 +1,49 @@
+    current_state = hass.states.get(entity_id).state
+    assert current_state
+    assert current_state == ValveState.OPEN
+    assert current_state == snapshot
```
It is checked.

zweckj commented on tests/components/watergate/test_init.py:
```diff
@@ -0,0 +1,72 @@
+    hass: HomeAssistant, mock_entry, mock_watergate_client
+) -> None:
+    """Test setting up the Watergate integration."""
+    hass.config.components.add("network")
```
```suggestion
```
I've never seen this in any other tests, so would be strange if you actually need this

zweckj commented on homeassistant/components/watergate/const.py:
```diff
@@ -0,0 +1,7 @@
+
+MANUFACTURER = "Watergate"
+
+VALVE_ENTITY_NAME = "valve"
```
preference, but since this is only used in `valve` I'd move it there

zweckj commented on homeassistant/components/watergate/strings.json:
```diff
@@ -0,0 +1,28 @@
+          "name": "[%key:common::config_flow::data::device%]",
+          "ip_address": "[%key:common::config_flow::data::ip%]"
+        },
+        "title": "Configure Watergate Device",
```
```suggestion
        "title": "Configure Watergate device",
```

zweckj commented on homeassistant/components/watergate/strings.json:
```diff
@@ -0,0 +1,28 @@
+          "name": "[%key:common::config_flow::data::device%]",
+          "ip_address": "[%key:common::config_flow::data::ip%]"
+        },
+        "title": "Reconfigure Watergate Device",
```
```suggestion
        "title": "Reconfigure Watergate device",
```

zweckj commented on tests/components/watergate/test_valve.py:
```diff
@@ -0,0 +1,45 @@
+"""Tests for the Watergate coordinator."""
```
```suggestion
"""Tests for the Watergate valve platform."""
```

zweckj commented on tests/components/watergate/test_valve.py:
```diff
@@ -0,0 +1,45 @@
+        blocking=True,
+    )
+
+    assert hass.states.get(entity_id).state == ValveState.CLOSING
```
```suggestion
    registered_entity = hass.states.get(entity_id)
    assert registered_entity
    assert registered_entity.state == ValveState.CLOSING
```

zweckj commented on tests/components/watergate/test_valve.py:
```diff
@@ -0,0 +1,45 @@
+        blocking=True,
+    )
+
+    assert hass.states.get(entity_id).state == ValveState.OPENING
```
```suggestion
    registered_entity = hass.states.get(entity_id)
    assert registered_entity
    assert registered_entity.state == ValveState.OPENING
```
that's what I meant with "assert state before". I'm surprised that mypy doesn't complain otherwise...

zweckj commented on tests/components/watergate/test_config_flow.py:
```diff
@@ -0,0 +1,89 @@
+    assert result["errors"][CONF_IP_ADDRESS] == "cannot_connect"
+
+    mock_watergate_client.async_get_device_state = AsyncMock(
+        return_value=DeviceState(
```
nitpicky, but duplicate code, use a variable

adam-the-hero replied:
Should I put it to conftest? I am missing this in documentation, what should goes where?

zweckj replied:
I normally put variables and helper functions in `init` and fixtures in conftest, so I'd say init

zweckj commented on homeassistant/components/watergate/__init__.py:
```diff
@@ -0,0 +1,101 @@
+
+    await coordinator.async_config_entry_first_refresh()
+
+    await watergate_client.async_set_webhook_url(
```
Oh I just noticed: Since you are never cleaning up the webhook, is this method idempotent?

adam-the-hero replied:
Yes, it is.

zweckj commented:
somehow can't see your recent commit to comment on, but two more things:

- the variable should be uppercase imo since it's a const
- for tests where you don't actually need the config entry to load (at least config flow & one of the init tests) you can patch away the entry creation for test speed
```python
@pytest.fixture
def mock_setup_entry() -> Generator[AsyncMock]:
    """Override async_setup_entry."""
    with patch(
        "homeassistant.components.watergate.async_setup_entry", return_value=True
    ) as mock_setup_entry:
        yield mock_setup_entry
```

adam-the-hero commented:
@zweckj I didn't see any performance increase, but I introduced this patching in init tests in inits, I probably need those. What are now next steps on this PR?

zweckj commented on tests/components/watergate/test_init.py:
```diff
@@ -0,0 +1,78 @@
+
+
+async def test_async_setup_entry(
+    hass: HomeAssistant, mock_entry, mock_watergate_client
```
```suggestion
    hass: HomeAssistant, mock_entry, mock_setup_entry: AsyncMock ,mock_watergate_client,
```
Here it should also be ok, plus please add type hints for all parameters in tests where it's missing

adam-the-hero replied:
Well, test is called test_async_setup_entry, and I am checking if webhooks are setup correctly so I cannot path the setup_entry.

zweckj replied:
Right brain lag. Don't forget adding the type hints though

zweckj commented:
> @zweckj I didn't see any performance increase, but I introduced this patching in init tests in inits, I probably need those. What are now next steps on this PR?

Once we are done (see my last review - I think that should be all from my side), I'll ask someone from the core team to review this, who will probably have some (hopefully minor) remarks and has the power to actually merge this.

zweckj commented:
> > > @zweckj I didn't see any performance increase, but I introduced this patching in init tests in inits, I probably need those. What are now next steps on this PR?
> 
> > 
> 
> > Once we are done (see my last review - I think that should be all from my side), I'll ask someone from the core team to review this, who will probably have some (hopefully minor) remarks and has the power to actually merge this.
> 
> 
> 
> So it will not be merged in this release?

Define this release. 2024.11 is cut already and in beta testing (and only will receive bugfixes, but no new features now), so 2024.12 is the next possible candidate.

adam-the-hero commented:
> > > > @zweckj I didn't see any performance increase, but I introduced this patching in init tests in inits, I probably need those. What are now next steps on this PR?
> > 
> > 
> > > 
> > 
> > 
> > > Once we are done (see my last review - I think that should be all from my side), I'll ask someone from the core team to review this, who will probably have some (hopefully minor) remarks and has the power to actually merge this.
> > 
> > 
> > So it will not be merged in this release?
> 
> Define this release. 2024.11 is cut already and in beta testing (and only will receive bug fixes, but no new features now), so 2024.12 is the next possible candidate.

Damn! :( That is not good news. I thought we would be able to deliver something in this release.

 So when it is merged, I can start creating PR for the rest of the Platforms and have everything introduced in one release, right? Since I loved the cooperation (even this nitpicky one), can I just assign a review to you? You know already a lot about what we are trying to deliver.

zweckj commented:
>  So when it is merged, I can start creating PR for the rest of the Platforms and have everything introduced in one release, right? 

Yes. Depends on your PRs but since we are at the start of the month chances are good. 

> Since I loved the cooperation (even this nitpicky one), can I just assign a review to you? You know already a lot about what we are trying to deliver.

You can - can't promise you anything though.

zweckj reviewed (approved):
LGTM 👍 Let's see what core team has to say

markvader committed:
Add Watergate Sonic Local Integration

joostlek commented on homeassistant/components/watergate/config_flow.py:
```diff
@@ -0,0 +1,49 @@
+
+WATERGATE_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_NAME): str,
```
Integrations are not allowed to have a settable name

joostlek commented on homeassistant/components/watergate/__init__.py:
```diff
@@ -0,0 +1,101 @@
+        async_generate_url(hass, webhook_id, allow_ip=True)
+    )
+
+    entry.runtime_data = WatergateData(watergate_client, coordinator, sonic_name)
```
You can reach the client via the `coordinator.api`, should we remove the dataclass wrapper and only put the coordinator in here?

joostlek commented on homeassistant/components/watergate/config_flow.py:
```diff
@@ -0,0 +1,49 @@
+            step_id="user", data_schema=WATERGATE_SCHEMA, errors=errors
+        )
+
+    async def validate_connection(self, ip_address: str):
+        """Validate the connection to the Watergate device."""
+        try:
+            return (
+                await WatergateLocalApiClient(ip_address).async_get_device_state()
+                is not None
+            )
+        except Exception:  # noqa: BLE001
+            return False
```
This can be inlined, it's small

adam-the-hero replied:
I changed it a little and I guess inlining will not be a solution here anymore.

joostlek commented on homeassistant/components/watergate/config_flow.py:
```diff
@@ -0,0 +1,49 @@
+                await WatergateLocalApiClient(ip_address).async_get_device_state()
+                is not None
+            )
+        except Exception:  # noqa: BLE001
```
Can we be more specific?

joostlek commented on homeassistant/components/watergate/entity.py:
```diff
@@ -0,0 +1,50 @@
+    ) -> None:
+        """Initialize the entity."""
+        super().__init__(coordinator)
+        self._attr_has_entity_name = True
```
Can be set outisde of the constructor

joostlek commented on homeassistant/components/watergate/entity.py:
```diff
@@ -0,0 +1,50 @@
+            identifiers={(DOMAIN, entry.entry_id)},
+            name=entry.data[CONF_NAME],
+            manufacturer=MANUFACTURER,
+            sw_version=coordinator.data.state.firmware_version
+            if coordinator.data.state
+            else "",
```
wrap multiline if statements in parenthesis

adam-the-hero replied:
It is not multiline anymore.

joostlek commented on homeassistant/components/watergate/entity.py:
```diff
@@ -0,0 +1,50 @@
+        super().__init__(coordinator)
+        self._attr_has_entity_name = True
+        self._api_client = coordinator.api
+        self._attr_unique_id = f"{entry.data[CONF_NAME]}.{entity_name}"
```
This is not a good unique id. It lacks unique ness

joostlek commented on homeassistant/components/watergate/entity.py:
```diff
@@ -0,0 +1,50 @@
+            if coordinator.data.state
+            else "",
+        )
+        self._api_client = entry.runtime_data.client
```
Why do we set the api client twice

adam-the-hero replied:
Good catch.

joostlek commented on homeassistant/components/watergate/entity.py:
```diff
@@ -0,0 +1,50 @@
+        self._api_client = entry.runtime_data.client
+
+
+@dataclass
+class WatergateData:
+    """Data for the Watergate integration."""
+
+    client: WatergateLocalApiClient
+    coordinator: WatergateDataCoordinator
+    sonic_name: str
```
Why does this belong here?

adam-the-hero replied:
It is required to create the entity's unique name, so it will be removed soon since I've already asked our engineers to introduce the Serial number in the device state response and I will use that plus the entity name.

joostlek replied:
No I mean, why does this class belong in this file

adam-the-hero replied:
If I remember correctly I had some circular dependencies and I thought that this is the best place since it is related to Entity. Where by default do you store those?

joostlek replied:
It's not related to the entity, rather to the integration on runtime, so I would propose coordinator.py or init.py. But in theory you don't even need this class anymore

joostlek commented on homeassistant/components/watergate/entity.py:
```diff
@@ -0,0 +1,50 @@
+    def __init__(
+        self,
+        coordinator: WatergateDataCoordinator,
+        entry: WatergateConfigEntry,
```
I always prefer not to have the entry in an entity as it feels like a big sluggish object

adam-the-hero replied:
Again since I am going to use the serial number which will be in the coordinator data, then I will just remove it.

joostlek commented on homeassistant/components/watergate/strings.json:
```diff
@@ -0,0 +1,28 @@
+        "title": "Configure Watergate device",
+        "description": "Please put name and IP address of your Watergate device"
+      },
+      "reconfigure": {
+        "data": {
+          "name": "[%key:common::config_flow::data::device%]",
+          "ip_address": "[%key:common::config_flow::data::ip%]"
+        },
+        "title": "Reconfigure Watergate device",
+        "description": "[%key:component::watergate::config::step::user::description%]"
+      }
```
unused

joostlek commented on homeassistant/components/watergate/strings.json:
```diff
@@ -0,0 +1,28 @@
+          "ip_address": "[%key:common::config_flow::data::ip%]"
+        },
+        "title": "Configure Watergate device",
+        "description": "Please put name and IP address of your Watergate device"
```
Please use data_description for the data fields instead of this

joostlek commented on homeassistant/components/watergate/strings.json:
```diff
@@ -0,0 +1,28 @@
+{
+  "config": {
+    "step": {
+      "confirm": {
+        "description": "[%key:common::config_flow::description::confirm_setup%]"
+      },
```
unused

adam-the-hero commented:
@joostlek I know that you have plenty of PR to be checked and I can wait, but I was wondering if I can just start implementing the rest of the Platforms based on this branch or should I wait because you will ask me to rewrite everything :)

adam-the-hero commented:
@joostlek @zweckj Any chances to put it forward? I know that our customers are already using custom cpomonent with that.

joostlek commented:
I'm currently busy with some refactoring of the quality scale, after that I can take a look again

adam-the-hero commented:
@joostlek Can someone else check it? Should I write a question regarding that?

adam-the-hero commented:
@joostlek @zweckj it is a cut off day very soon. Can we find someone who can do the review? I hope that this is not much to be done here.

zweckj commented on tests/components/watergate/__init__.py:
```diff
@@ -0,0 +1,25 @@
+
+from homeassistant.core import HomeAssistant
+
+DEFAULT_DEVICE_STATE = DeviceState(
```
move two const

zweckj commented on tests/components/watergate/test_valve.py:
```diff
@@ -0,0 +1,57 @@
+from tests.common import AsyncMock, MockConfigEntry
+
+
+async def test_change_valve_state(
```
split this into two tests: one for snapshots, one for actions

zweckj commented on homeassistant/components/watergate/entity.py:
```diff
@@ -0,0 +1,30 @@
+            name="Sonic",
+            serial_number=coordinator.data.serial_number,
+            manufacturer=MANUFACTURER,
+            sw_version=coordinator.data.firmware_version if coordinator.data else "",
```
```suggestion
            sw_version=coordinator.data.firmware_version if coordinator.data else None,
```
better to leave it default it can't be set

zweckj commented on homeassistant/components/watergate/config_flow.py:
```diff
@@ -0,0 +1,57 @@
+            step_id="user", data_schema=WATERGATE_SCHEMA, errors=errors
+        )
+
+    async def validate_connection(self, ip_address: str):
```
I think this would look cleaner if we'd inline this

adam-the-hero replied:
I've changed this logic a little to make sure we are using http. So it will not be as beautiful when inlined.

zweckj reviewed (commented):
I can't do anything at this point and joost already said he's busy. I found a couple more things though (some which I only learned recently)

zweckj commented:
The requirements have also changed in November, you'll need to add a `quality_scale.yaml` now, fill it out and reach at least `Bronze` https://developers.home-assistant.io/docs/core/integration-quality-scale/

joostlek reviewed (changes requested):
As Josef said we now require a quality_scale.yaml in the integration. I do think that you can check of a lot of boxes for that. So it shouldn't be too big and we can merge this quite quick

adam-the-hero commented on homeassistant/components/watergate/__init__.py:
```diff
@@ -0,0 +1,99 @@
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: WatergateConfigEntry) -> bool:
+    """Unload a config entry."""
+    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
```
Moved. I will test how it works.

adam-the-hero commented on homeassistant/components/watergate/__init__.py:
```diff
@@ -0,0 +1,99 @@
+    coordinator = WatergateDataCoordinator(hass, watergate_client)
+    entry.runtime_data = coordinator
+
+    hass.components.webhook.async_unregister(webhook_id)
```
I had an issues with registering the webhook in next line after reconfiguring or restarting HA. It looked like webhook.async_register is not idempotent. I will check it again.

adam-the-hero commented on tests/components/watergate/test_config_flow.py:
```diff
@@ -0,0 +1,84 @@
+
+@pytest.mark.parametrize(
+    "client_result",
+    [AsyncMock(return_value=None), AsyncMock(side_effect=WatergateApiException)],
```
When there is no data.

adam-the-hero commented:
I have done everything, but I cannot commit due to:
``` hassfest.................................................................Failed
- hook id: hassfest
- exit code: 1

Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/workspaces/core/script/hassfest/__main__.py", line 11, in <module>
    from . import (
  File "/workspaces/core/script/hassfest/docker.py", line 7, in <module>
    from homeassistant.components.go2rtc.const import RECOMMENDED_VERSION as GO2RTC_VERSION
  File "/workspaces/core/homeassistant/components/go2rtc/__init__.py", line 8, in <module>
    from go2rtc_client import Go2RtcRestClient
ModuleNotFoundError: No module named 'go2rtc_client'
```

I've rectreated the whole Dev COntainer but still something is messy there.
Do you know what is it about?

adam-the-hero commented:
I have changed everything but I am not able to commit.
I am getting such error:
```
hassfest.................................................................Failed
- hook id: hassfest
- exit code: 1

Error loading libturbojpeg; Camera snapshot performance will be sub-optimal
Traceback (most recent call last):
  File "/workspaces/core/homeassistant/components/camera/img_util.py", line 100, in __init__
    TurboJPEGSingleton.__instance = TurboJPEG()
                                    ^^^^^^^^^
NameError: name 'TurboJPEG' is not defined
Warning: homeassistant/components/dte_energy_bridge/__init__.py and manifest.json missing, skipping directory. If this is your development environment, you can safely delete this folder.
Validating application_credentials... done in 0.09s
Validating bluetooth... done in 0.01s
Validating codeowners... done in 0.01s
Validating config_schema... done in 0.65s
Validating dependencies... done in 1.81s
Validating dhcp... done in 0.01s
Validating icons... done in 0.21s
Validating json... done in 0.00s
Validating manifest... done in 0.02s
Validating mqtt... done in 0.01s
Validating quality_scale... done in 0.12s
Validating requirements... done in 0.01s
Validating services... done in 0.61s
Validating ssdp... done in 0.01s
Validating translations... done in 3.49s
Validating usb... done in 0.01s
Validating zeroconf... done in 0.01s
Validating config_flow... done in 0.06s
Validating docker... done in 0.00s
Validating mypy_config... done in 0.04s
Validating metadata... done in 0.00s

Integrations: 1304
Invalid integrations: 1

Found errors. Generating files canceled.

Integration assist_satellite:
* [ERROR] [SERVICES] Invalid services.yaml: Unknown supported feature 'assist_satellite.AssistSatelliteEntityFeature.ANNOUNCE' @ data['announce']['target']['entity'][0]['supported_features'][0]. Got None
```
I've rebased to upstream/dev, Dev container is recreated and still there are some issues.

joostlek commented:
Please use the ready for review button when ready for review. Please don't forget to set the quality_scale to bronze in the manifest

joostlek commented:
The CI is failing, can you take a look?

adam-the-hero commented:
I've checked the failed tests, but those failures are related to other components, which I didn't event touch. Is it normal or I really changed something that makes those failures?

zweckj commented:
> Is it normal or I really changed something that makes those failures?

This happens sometimes and you can ignore that if it's not related to your component.

joostlek commented on tests/components/watergate/test_config_flow.py:
```diff
@@ -0,0 +1,83 @@
+
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+    assert result["title"] == "Sonic"
+    assert result["data"] == {**user_input, CONF_WEBHOOK_ID: MOCK_WEBHOOK_ID}
```
Let's also assert the unique id here
`assert result["result"].unique_id == ...` I believe

joostlek commented on tests/components/watergate/conftest.py:
```diff
@@ -0,0 +1,71 @@
+
+
+@pytest.fixture
+def mock_entry() -> MockConfigEntry:
```
This config entry is missing an unique_id

joostlek commented on tests/components/watergate/test_config_flow.py:
I am missing a test where we add the mockconfigentry and try to add the same instance again. That should abort since the entry is already known

joostlek commented on tests/components/watergate/test_config_flow.py:
```diff
@@ -0,0 +1,127 @@
+    assert result["title"] == "Sonic"
+    assert result["data"] == {**user_input, CONF_WEBHOOK_ID: MOCK_WEBHOOK_ID}
+
+    entries = hass.config_entries.async_entries(DOMAIN)
+    assert entries[0].unique_id == DEFAULT_SERIAL_NUMBER
```
```suggestion
    assert result["result"].unique_id == DEFAULT_SERIAL_NUMBER
```

joostlek commented on tests/components/watergate/test_config_flow.py:
```diff
@@ -0,0 +1,127 @@
+
+async def test_abort_if_id_is_not_unique(
+    hass: HomeAssistant,
+    mock_watergate_client: Generator[AsyncMock],
+    mock_webhook_id_generation: Generator[None],
+    mock_setup_entry: Generator[AsyncMock],
+    user_input: dict[str, str],
+) -> None:
+    """Test checking if we will inform user that this entity is already registered."""
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_USER}
+    )
+
+    assert result["type"] is FlowResultType.FORM
+    assert result["step_id"] == "user"
+    assert CONF_IP_ADDRESS in result["data_schema"].schema
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"], user_input
+    )
+
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+    assert result["title"] == "Sonic"
+    assert result["data"] == {**user_input, CONF_WEBHOOK_ID: MOCK_WEBHOOK_ID}
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_USER}
+    )
+
+    assert result["type"] is FlowResultType.FORM
+    assert result["step_id"] == "user"
+    assert CONF_IP_ADDRESS in result["data_schema"].schema
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"], user_input
+    )
+
+    assert result["type"] is FlowResultType.ABORT
+    assert result["reason"] == "already_configured"
```
Please use `mock_entry` for this. you just have to add it to hass and then try to set it up

joostlek commented on homeassistant/components/watergate/quality_scale.yaml:
```diff
@@ -0,0 +1,43 @@
+  common-modules: done
+  config-flow-test-coverage: done
+  config-flow: done
+  dependency-transparency: done
```
Looking at the repository https://github.com/watergate-ai/watergate-local-api-python, we don't build and publish the dependency in the CI. That is required to increase transparency and security. For example, the integration uses 2024.2.1, while the repository only has 2024.1.11.

adam-the-hero replied:
Done!

joostlek commented:
After this I think this PR is ready to merge

joostlek commented:
@adam-the-hero If you use discord, can you please send me a message?
